server_code/historic_server/persistence.py

```diff
@@ -98,11 +107,11 @@ def _record_event(event, prevent_duplication):
     prevent_duplication : bool
         Whether to disallow records where the state is unchanged from previously
     """
-    if event.event_type == "creation" and event.affected.uid is not None:
-        raise AttributeError("Object uid cannot be assigned on creation")
-
     if event.event_type == "creation":
-        event.affected.uid = uuid4().hex
+        if event.affected.uid is None:
+            event.affected.uid = uuid4().hex
+        elif not _is_valid_uid(event.affected.uid):
+            raise InvalidUIDError(f"Invalid UID {event.affected.uid}")
```

Comment:
Good point. Probably not needed.

Reply:
Actually, I've left this alone for now so that we have a portable exception. If I try using a ValueError, it doesn't get passed back to the client nicely.